posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+category: 'Metaprogramming'
+date: 2025-03-29 12:33:00 UTC+00:00
+description: 'Brief introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: what-is-a-simproc
+tags: 'simp, simproc, meta'
+title: What is a simproc?
+type: text
+---
+
+Lean core recently added support for custom simplification procedures, aka *simprocs*.
+This blog post aims to explain what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+The first part of this post is a purely informal description of what simprocs are and do.
+The second part is a walkthrough to writing a simple simproc in three different ways.
+
+# What is a simproc
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.
+
+## How simp works
+
+On an informal level, `simp` can be seen as recursively traversing the expression `e` to be simplified from the outside-in, each time checking whether the expression matches the left hand side `LHS` of some simp lemma of the form `LHS = RHS`, and replace `LHS` by `RHS` in the expression if it does match.
+
+To see what steps `simp` performs, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.
+
+Here is an example of a series of rewrites performed by `simp` on a simple goal:
+```lean
+set_option trace.Meta.Tactic.simp true in
+example : 37 * (Nat.fib 0 + 0) = 0 := by
+  simp
+```
+which yields
+```
+[Meta.Tactic.simp.rewrite] Nat.fib_zero:1000:
+      Nat.fib 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] add_zero:1000:
+      0 + 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] mul_zero:1000:
+      37 * 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] eq_self:1000:
+      0 = 0
+    ==>
+      True
+```
+i.e. the steps are:
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+In this picture, simp lemmas are *fixed* rules to turn a *specific* left hand side into a *specific* right hand side.
+In contrast, simprocs are *flexible* rules to turn a *specific* left hand side into a right hand side *computed* from the left hand side.
+In this sense, they are *parametric simp lemmas*.
+
+## Examples of simprocs
+
+In this section, we exemplify three simprocs through the following use cases:
+* Avoiding combinatorial explosion of lemmas
+* Performance optimisation
+* Computation
+
+### Avoiding combinatorial explosion of lemmas: The `existsAndEq` simproc
+
+The `existsAndEq` simproc is designed to simplify expressions of the form `∃ a, ... ∧ a = a' ∧ ...` where `a'` is some quantity independent of `a'` by removing the existential quantifier and replacing all occurences of `a` by `a'`.
+
+```lean
+example : ∃ (a : ℤ), a*a = 25 ∧ a = 5 := by
+  simp +arith only [existsAndEq, and_true]
```

Comment:
`+arith only` is (in my opinion) confusing (by which I mean: I have no idea what this does, and this is supposed to be the very first example). Can't you just run `simp` on the example and show what it turns it into? Also I am confused about why you need `and_true` given your claim about what `existsAndEq` does. Surely your argument claims that `simp [existsAndEq]` turns the goal into `5*5 = 25`? Oh -- it replaces `a = 5` by `5 = 5`?? Maybe you should explain all this clearly rather than making the user have to work it all out? I think it would be much clearer just to show the output after `simp [existsAndEq]` for pedagogical purposes rather than insisting on closing the goal.

Reply:
Or perhaps after `simp only [existsAndEq]`